File: 백준/Gold/2573. 빙산/빙산.py
import sys
input = lambda : sys.stdin.readline().strip()

n,m=map(int,input().split())
space=[list(map(int,input().split())) for i in range(n)]
ices = [(i,j) for i in range(n) for j in range(m) if space[i][j]]

direction = [[0,1],[1,0],[0,-1],[-1,0]]
visited=[[1]*m for i in range(n)]
time=0
flag=True
while ices:
    if flag:
        x,y = ices[0]
        visited[x][y]=time
        queue = [[x,y]]
        cnt=1
        while queue:
            i,j=queue.pop()
            for dx, dy in direction:
                x,y = i+dx,j+dy
                if 0<=x<n and 0<=y<m and space[x][y] and visited[x][y]!=time:
                    visited[x][y]=time
                    queue.append([x,y])
                    cnt+=1
        if cnt < len(ices):
            print(time*-1)
            break
    time-=1
    flag=False            
    
    
    
    diff = []
    for ice in ices:
        cnt=0
        for dx, dy in direction:
            x,y = ice[0]+dx,ice[1]+dy
            if 0<=x<n and 0<=y<m and space[x][y]==0:
                cnt+=1
        diff.append(cnt)
    new=[]
    for index in range(len(ices)):
        i,j = ices[index]
        space[i][j] = max(0, space[i][j]-diff[index])
        index+=1
        if space[i][j]:
            new.append((i,j))
        else:
            flag=True
    ices=new
else:
    print(0)

File: 백준/Gold/2573. 빙산/빙산.java
import java.io.*;
import java.util.*;

public class Main {
    static int n, m;
    static int[][] space;
    static List<int[]> ices = new ArrayList<>();
    static int[][] visited;
    static final int[][] direction = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        // 입력 처리
        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());

        space = new int[n][m];
        visited = new int[n][m];

        for (int i = 0; i < n; i++) {
            Arrays.fill(visited[i], 1);  // visited 초기값 1로 설정 (파이썬과 동일)
        }

        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < m; j++) {
                space[i][j] = Integer.parseInt(st.nextToken());
                if (space[i][j] > 0) {
                    ices.add(new int[]{i, j});
                }
            }
        }

        int time = 0;
        boolean flag = true;

        while (!ices.isEmpty()) {
            if (flag) {
                int[] start = ices.get(0);
                int x = start[0], y = start[1];
                visited[x][y] = time;
                List<int[]> queue = new ArrayList<>();
                queue.add(new int[]{x, y});
                int cnt = 1;

                while (!queue.isEmpty()) {
                    int[] cur = queue.remove(queue.size() - 1);  // stack pop
                    int i = cur[0], j = cur[1];

                    for (int[] d : direction) {
                        int nx = i + d[0];
                        int ny = j + d[1];

                        if (nx >= 0 && nx < n && ny >= 0 && ny < m &&
                            space[nx][ny] > 0 && visited[nx][ny] != time) {
                            visited[nx][ny] = time;
                            queue.add(new int[]{nx, ny});
                            cnt++;
                        }
                    }
                }

                if (cnt < ices.size()) {
                    System.out.println(-time);
                    return;
                }
            }

            time--;
            flag = false;

            List<Integer> diff = new ArrayList<>();
            for (int[] ice : ices) {
                int i = ice[0], j = ice[1];
                int water = 0;

                for (int[] d : direction) {
                    int x = i + d[0];
                    int y = j + d[1];

                    if (x >= 0 && x < n && y >= 0 && y < m && space[x][y] == 0) {
                        water++;
                    }
                }

                diff.add(water);
            }

            List<int[]> newIces = new ArrayList<>();
            for (int index = 0; index < ices.size(); index++) {
                int[] ice = ices.get(index);
                int i = ice[0], j = ice[1];
                space[i][j] = Math.max(0, space[i][j] - diff.get(index));

                if (space[i][j] > 0) {
                    newIces.add(new int[]{i, j});
                } else {
                    flag = true;
                }
            }

            ices = newIces;
        }

        System.out.println(0);
    }
}
